File: lib/android/src/main/java/com/reactnativeldk/classes/LdkFeeEstimator.kt
package com.reactnativeldk.classes
import com.reactnativeldk.EventTypes
import com.reactnativeldk.LdkEventEmitter
import org.ldk.enums.ConfirmationTarget
import org.ldk.structs.FeeEstimator

class LdkFeeEstimator {
    var anchorChannelFee: Int = 0
    var nonAnchorChannelFee: Int = 0
    var channelCloseMinimum: Int = 0
    var minAllowedAnchorChannelRemoteFee: Int = 0
    var minAllowedNonAnchorChannelRemoteFee: Int = 0
    var outputSpendingFee: Int = 0
    var maximumFeeEstimate: Int = 0
    var urgentOnChainSweep: Int = 0

    fun update(
        anchorChannelFee: Int,
        nonAnchorChannelFee: Int,
        channelCloseMinimum: Int,
        minAllowedAnchorChannelRemoteFee: Int,
        minAllowedNonAnchorChannelRemoteFee: Int,
        outputSpendingFee: Int,
        maximumFeeEstimate: Int,
        urgentOnChainSweep: Int
    ) {
        this.anchorChannelFee = anchorChannelFee
        this.nonAnchorChannelFee = nonAnchorChannelFee
        this.channelCloseMinimum = channelCloseMinimum
        this.minAllowedAnchorChannelRemoteFee = minAllowedAnchorChannelRemoteFee
        this.minAllowedNonAnchorChannelRemoteFee = minAllowedNonAnchorChannelRemoteFee
        this.outputSpendingFee = outputSpendingFee
        this.maximumFeeEstimate = maximumFeeEstimate
        this.urgentOnChainSweep = urgentOnChainSweep

        LdkEventEmitter.send(EventTypes.native_log, "Fee estimator updated")
    }

    var feeEstimator = FeeEstimator.new_impl { target: ConfirmationTarget ->
        return@new_impl when (target) {
            ConfirmationTarget.LDKConfirmationTarget_AnchorChannelFee -> anchorChannelFee
            ConfirmationTarget.LDKConfirmationTarget_NonAnchorChannelFee -> nonAnchorChannelFee
            ConfirmationTarget.LDKConfirmationTarget_ChannelCloseMinimum -> channelCloseMinimum
            ConfirmationTarget.LDKConfirmationTarget_MinAllowedAnchorChannelRemoteFee -> minAllowedAnchorChannelRemoteFee
            ConfirmationTarget.LDKConfirmationTarget_MinAllowedNonAnchorChannelRemoteFee -> minAllowedNonAnchorChannelRemoteFee
            ConfirmationTarget.LDKConfirmationTarget_OutputSpendingFee -> outputSpendingFee
            ConfirmationTarget.LDKConfirmationTarget_MaximumFeeEstimate -> maximumFeeEstimate
            ConfirmationTarget.LDKConfirmationTarget_UrgentOnChainSweep -> urgentOnChainSweep
        }
    }
}

File: lib/ios/Classes/LdkFeeEstimator.swift
//
//  LdkfeeEstimator.swift
//  react-native-ldk
//
//  Created by Jason van den Berg on 2022/05/10.
//

import Foundation
import LightningDevKit

class LdkFeeEstimator: FeeEstimator {
    private var anchorChannelFee: UInt32 = 0
    private var nonAnchorChannelFee: UInt32 = 0
    private var channelCloseMinimum: UInt32 = 0
    private var minAllowedAnchorChannelRemoteFee: UInt32 = 0
    private var minAllowedNonAnchorChannelRemoteFee: UInt32 = 0
    private var outputSpendingFee: UInt32 = 0
    private var maximumFeeEstimate: UInt32 = 0
    private var urgentOnChainSweep: UInt32 = 0

    func update(
        anchorChannelFee: UInt32,
        nonAnchorChannelFee: UInt32,
        channelCloseMinimum: UInt32,
        minAllowedAnchorChannelRemoteFee: UInt32,
        minAllowedNonAnchorChannelRemoteFee: UInt32,
        outputSpendingFee: UInt32,
        maximumFeeEstimate: UInt32,
        urgentOnChainSweep: UInt32
    ) {
        self.anchorChannelFee = anchorChannelFee
        self.nonAnchorChannelFee = nonAnchorChannelFee
        self.channelCloseMinimum = channelCloseMinimum
        self.minAllowedAnchorChannelRemoteFee = minAllowedAnchorChannelRemoteFee
        self.minAllowedNonAnchorChannelRemoteFee = minAllowedNonAnchorChannelRemoteFee
        self.outputSpendingFee = outputSpendingFee
        self.maximumFeeEstimate = maximumFeeEstimate
        self.urgentOnChainSweep = urgentOnChainSweep

        LdkEventEmitter.shared.send(withEvent: .native_log, body: "Fee estimator updated")
    }

    override func getEstSatPer1000Weight(confirmationTarget: Bindings.ConfirmationTarget) -> UInt32 {
        let target = confirmationTarget

        switch target {
        case .AnchorChannelFee:
            return anchorChannelFee
        case .NonAnchorChannelFee:
            return nonAnchorChannelFee
        case .ChannelCloseMinimum:
            return channelCloseMinimum
        case .MinAllowedAnchorChannelRemoteFee:
            return minAllowedAnchorChannelRemoteFee
        case .MinAllowedNonAnchorChannelRemoteFee:
            return minAllowedNonAnchorChannelRemoteFee
        case .OutputSpendingFee:
            return outputSpendingFee
        case .MaximumFeeEstimate:
            return maximumFeeEstimate
        case .UrgentOnChainSweep:
            return urgentOnChainSweep
        }
    }
}
